docs/fomega/mutual-type-level-recursion/MutualData.agda

```diff
@@ -0,0 +1,420 @@
+  -- In this module we define the `TreeForest` data type using `IFix`.
+  -- The type of tags is as before:
+
+  data TreeForestᵗ : Set where
+    Treeᵗ Forestᵗ : TreeForestᵗ
+
+  -- The definition of `TreeForest` is similar to the definition from the previous section:
+  -- we again have the `Tree` and `Forest` type aliases, they're just local now,
+  -- and as before we dispatch on `tfᵗ` and describe the constructors of the `Tree` and `Forest`
+  -- data types using the sum-of-products representation:
+
+  TreeForest : Set -> TreeForestᵗ -> Set
+  TreeForest A =
+    IFix λ Rec tfᵗ ->
+      let Tree   = Rec Treeᵗ
+          Forest = Rec Forestᵗ
+        in case tfᵗ of λ
+          { Treeᵗ   -> A × Forest
+          ; Forestᵗ -> ⊤ ⊎ Tree × Forest
+          }
+
+  -- We've cheated a little bit here by keeping `A` outside of the pattern functor.
+  -- That's just for simplicity, there is no fundamental limitation that prevents us
+  -- from taking fixed points of pattern functors of arbitrary arity.
+  -- Which is a topic for another document.
+
+  Tree : Set -> Set
+  Tree A = TreeForest A Treeᵗ
+
+  Forest : Set -> Set
+  Forest A = TreeForest A Forestᵗ
+
+  pattern node x forest    = wrap (x , forest)
+  pattern nil              = wrap (inj₁ tt)
+  pattern cons tree forest = wrap (inj₂ (tree , forest))
+
+  -- The sanity check.
+
+  mutual
+    toTree₀ : ∀ {A} -> Tree A -> Tree₀ A
+    toTree₀ (node x forest) = node₀ x (toForest₀ forest)
+
+    toForest₀ : ∀ {A} -> Forest A -> Forest₀ A
+    toForest₀  nil               = nil₀
+    toForest₀ (cons tree forest) = cons₀ (toTree₀ tree) (toForest₀ forest)
+
+module JoinedComputeIFixFω where
+  -- The types we had previously are not System Fω compatible, because we matched on tags at the
+  -- type level. Can we avoid this? Yes, easily. Instead of defining data types representing tags
+  -- we can just lambda-encode them, because we do have lambdas at the type level in System Fω.
+
+  -- The Church-encoded version of `TreeForestᵗ` at the type level looks like this:
+
+  -- ∀ R -> (Set -> Set -> R) -> R
+
+  -- However this is still not System Fω compatible, because here we quantify over `R` which in
+  -- System Fω is a kind and we can't quatify over kinds there. But looking at
+
+  -- TreeForestF : Set -> TreeForestᵗ -> Set
+  -- TreeForestF A Treeᵗ   = A × Forest A
+  -- TreeForestF A Forestᵗ = ⊤ ⊎ Tree A × Forest A
+
+  -- we see that we match on a `TreeForestᵗ` only to return a `Set`. Hence we can just instantiate
+  -- `R` to `Set` right away as we do not do anything else with tags. We arrive at
+
+  TreeForestᵗ : Set₁
+  TreeForestᵗ = Set -> Set -> Set
+
+  Treeᵗ : TreeForestᵗ
+  Treeᵗ A B = A
+
+  Forestᵗ : TreeForestᵗ
+  Forestᵗ A B = B
+
+  -- Where `TreeForestᵗ` is the type of inlined matchers which always return a `Set` and
+  -- `Treeᵗ` and `Forestᵗ` are such matchers. The rest is straightforward:
+
+  TreeForest : Set -> TreeForestᵗ -> Set
+  TreeForest A =
+    IFix λ R tfᵗ ->
```

Comment:
Can we call this `Rec` just for consistency?

Reply:
Err, wanted to replace `R` by `Rec` everywhere and forgot. Thanks.